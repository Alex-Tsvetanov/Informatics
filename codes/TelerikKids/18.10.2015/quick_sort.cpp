#include <iostream>
#include <random>

using namespace std;

template <typename T>
T* get_pivot (T* a, T* b)
{
	T* masterPivot = a;
	T averrage = 0;
	for (T* it = a ; it != b ; it ++)
		averrage += (*it);
	averrage /= (b - a);
	for (T* it = a ; it != b ; it ++)
		if (abs ((*masterPivot) - averrage) > abs ((*it) - averrage))
			masterPivot = it;

	return masterPivot;
}

template <typename T>
T* fillArr (T* newArr, T* pivot, T* s, T* e)
{
	T* indOfPivot;
	size_t ind = 0;
	for (T* it = s ; it != e ; it ++)
		if ((*it) < (*pivot))
		{
			newArr [ind] = *it;
			ind ++;
		}
	indOfPivot = newArr + ind;
	for (T* it = s ; it != e ; it ++)
		if ((*it) >= (*pivot))
		{
			newArr [ind] = *it;
			ind ++;
		}
	return indOfPivot;
}

template <typename T>
void copyArr (T* newArr, T* s, T* e)
{
	T* newArrIt = newArr;
	for (T* it = s ; it != e ; it ++, newArrIt ++)
		*it = *newArrIt;
}

template <typename T>
void quick_sort (T* s, T* e)
{
	if (s + 1 == e)
		return ;
	if (s == e)
		return ;
	T* newArr = new T [e - s];
	T* pivot = fillArr (newArr, get_pivot (s, e), s, e);
	//for (T* it = s ; it != e ; it ++)
	//	cout << *it << " ";
	//cout << "\n" << *pivot << "\n----------------------------" << endl;
	quick_sort (newArr, pivot + 1);
	quick_sort (pivot + 1, newArr + (e - s));
	copyArr (newArr, s, e);
	delete[] newArr;
}

int main ()
{
	size_t SIZE;
	cin >> SIZE;
	int* arr = new int [SIZE];
	for (size_t i = 0 ; i < SIZE ; i ++)
		cin >> arr [i];
	quick_sort (arr + 0, arr + SIZE);
	for (size_t i = 0 ; i < SIZE ; i ++)
		cout << arr [i] << " ";
	cout << "\n";
}
