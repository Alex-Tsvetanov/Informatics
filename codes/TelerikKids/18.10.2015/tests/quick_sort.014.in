59
40 54 13 1 93 3 68 61 2 90 28 54 46 64 28 31 38 71 45 62 32 48 23 40 6 4 49 35 3 75 40 79 82 61 37 56 73 43 41 42 72 97 94 47 57 87 18 59 13 55 5 94 79 69 92 10 28 27 50 
