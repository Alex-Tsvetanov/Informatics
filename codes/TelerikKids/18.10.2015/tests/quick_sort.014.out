1 2 3 3 4 5 6 10 13 13 18 23 27 28 28 28 31 32 35 37 38 40 40 40 41 42 43 45 46 47 48 49 50 54 54 55 56 57 59 61 61 62 64 68 69 71 72 73 75 79 79 82 87 90 92 93 94 94 97 
